nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -818,35 +822,57 @@ impl DataStore {
+        // at this point in query construction, because here, the selection
+        // contains both `sled_dsl::id` and `vmm_dsl::id` columns, but it does
+        // *not* have anything that makes it no longer implement
+        // `diesel::QuerySource`, which the `paginated_multicolumn` function
+        // requires. This ordering doesn't actually matter when it comes to the
+        // generated SQL, which should be equivalent no matter how we construct
+        // the query, but it *does* matter for satisfying Diesel's trait
+        // constraints.
+        let query = paginated_multicolumn(
+            query,
+            (sled_dsl::id, vmm_dsl::id),
+            pagparams,
+        );
+        // Filter out sleds that aren't in service, and VMMs that aren't
+        // incarnated on a sled.
+        let query = query
             .filter(sled_dsl::time_deleted.is_null())
             .sled_filter(SledFilter::InService)
-            .inner_join(
-                vmm_dsl::vmm
-                    .on(vmm_dsl::sled_id
-                        .eq(sled_dsl::id)
-                        .and(vmm_dsl::time_deleted.is_null())
-                        // Ignore instances which are in states that are not
-                        // known to exist on a sled. Since this query drives
-                        // instance-watcher health checking, it is not necessary
-                        // to perform health checks for VMMs that don't actually
-                        // exist in real life.
-                        .and(
-                            vmm_dsl::state.ne_all(VmmState::NONEXISTENT_STATES),
-                        ))
-                    .inner_join(
-                        instance_dsl::instance
-                            .on(instance_dsl::id
-                                .eq(vmm_dsl::instance_id)
-                                .and(instance_dsl::time_deleted.is_null()))
-                            .inner_join(
-                                project_dsl::project.on(project_dsl::id
-                                    .eq(instance_dsl::project_id)
-                                    .and(project_dsl::time_deleted.is_null())),
-                            ),
-                    ),
-            )
-            .sled_filter(SledFilter::InService)
+            .filter(vmm_dsl::time_deleted.is_null())
+            // Ignore VMMs which are in states that are not known to exist on a
+            // sled. Since this query drives instance-watcher health checking,
+            // it is not necessary to perform health checks for VMMs that don't
+            // actually exist in real life.
+            .filter(vmm_dsl::state.ne_all(VmmState::NONEXISTENT_STATES));
+        // Now, join with the `instance` table on the instance's VMM ID.
+        let query = query.inner_join(
+            instance_dsl::instance.on(instance_dsl::id
+                .eq(vmm_dsl::instance_id)
+                .and(instance_dsl::time_deleted.is_null())),
```

Comment:
Similarly, what would mean if a row matched everything thus far but the instance or project was deleted?

Reply:
hmm...i think a deleted instance or project still having active VMMs is _probably_ a bug? but, i think it would be reasonable to change this query to not care if the instance or project is deleted, regardless. I'll just do that.

Reply:
I removed those filters in https://github.com/oxidecomputer/omicron/pull/6534/commits/6f400293be7a783f7a95a2c6ae3ef06daeb1d73d.